// ortoni-report/dist/utils/utils.js
import path from "path";
export function normalizeFilePath(filePath) {
    // Normalize the path to handle different separators
    const normalizedPath = path.normalize(filePath);
    // Get the base name of the file (removes any leading directories)
    return path.basename(normalizedPath);
}
export function formatDate(date) {
    const day = String(date.getDate()).padStart(2, "0");
    const month = date.toLocaleString("default", { month: "short" });
    const year = date.getFullYear();
    const time = date.toLocaleTimeString();
    return `${day}-${month}-${year} ${time}`;
}
export function safeStringify(obj, indent = 2) {
    const cache = new Set();
    const json = JSON.stringify(obj, (key, value) => {
        if (typeof value === "object" && value !== null) {
            if (cache.has(value)) {
                return;
            }
            cache.add(value);
        }
        return value;
    }, indent);
    cache.clear();
    return json;
}
export function ensureHtmlExtension(filename) {
    const ext = path.extname(filename);
    if (ext && ext.toLowerCase() === ".html") {
        return filename;
    }
    return `${filename}.html`;
}
export function escapeHtml(unsafe) {
    if (typeof unsafe !== "string") {
        return String(unsafe);
    }
    return unsafe.replace(/[&<"']/g, function (match) {
        const escapeMap = {
            "&": "&amp;",
            "<": "&lt;",
            ">": "&gt;",
            '"': "&quot;",
            "'": "&#039;",
        };
        return escapeMap[match] || match;
    });
}
export function formatDateUTC(date) {
    return date.toISOString();
}
export function formatDateLocal(dateInput) {
    if (!dateInput)
        return "N/A";
    try {
        const date = typeof dateInput === "string" ? new Date(dateInput) : dateInput;
        if (isNaN(date.getTime())) {
            return "N/A";
        }
        const options = {
            year: "numeric",
            month: "short",
            day: "2-digit",
            hour: "2-digit",
            minute: "2-digit",
            hour12: true,
        };
        return new Intl.DateTimeFormat("en-US", options).format(date);
    }
    catch (e) {
        return "N/A";
    }
}
export function formatDateNoTimezone(isoString) {
    const date = new Date(isoString);
    return date.toLocaleString("en-US", {
        dateStyle: "medium",
        timeStyle: "short",
    });
}
export function extractSuites(titlePath) {
    const tagPattern = /@[\w]+/g;
    const suiteParts = titlePath
        .slice(3, titlePath.length - 1)
        .map((p) => p.replace(tagPattern, "").trim());
    return {
        hierarchy: suiteParts.join(" > "),
        topLevelSuite: suiteParts[0] ?? "",
        parentSuite: suiteParts[suiteParts.length - 1] ?? "", // last suite
    };
}
